tpl: Add localizedDate template function

soundrussian opened the pull request:
This adds a template function to implement date localization, as was discussed in issue #245. All credit should really go to @variadico , I've just created a pull request to learn myself some Go.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/spf13/hugo?pullRequest=2041) <br/>You should sign our Contributor License Agreement in order to get your pull request merged.<br/>

digitalcraftsman commented:
Hi @soundrussian,

it's a good idea to implement such a template function, especially while looking at #1744. 

> The default locale is based on [environment variables]. However, if people want to change the locale, they can start Hugo like this.

We should find a workaround once we have i18n support as a feature. Inside the template we will have access to the locale of the current page. Depending on the locale visitors should see the date formatted respectively. But this seems only to be possible when we set / change the environment accordingly.

[environment variables]: https://godoc.org/github.com/variadico/lctime

variadico commented:
Actually, it is possible to set the locale during runtime by calling `lctime.SetLocale`.

```go
lctime.SetLocale("es_MX")
fmt.Println(lctime.Strftime("%c", time.Now()))
// prints: lun 14 dic 2015 22:31:56 PST
```
From the [readme](https://github.com/variadico/lctime).

digitalcraftsman commented:
My bad. Thanks for the correction. Anyhow, it's a powerful combination.

bep commented:
This one needs some thinking, the bigger picture and all.

soundrussian commented:
I've updated the code to set locale from `languageCode` instead of environment variable, this seems more user-friendly. There might be some differences though between the formats of `languageCode` and expected locale, which I tried to address, but I'm not sure how `languageCode` is used in the wild.

digitalcraftsman commented:
Could you also add a few words for the documentation? Take a look at the examples in `docs/content/templates/functions.md`.

Maybe it seems trivial but why I would use `localizeDate` instead of <code>localize<b>d</b>Date</code> as function name.

soundrussian commented:
I agree that `localizeDate` is a better name, since it implies some action on the provided arguments. Changed the name and added an extensive documentation.

digitalcraftsman commented:
> bep: This one needs some thinking, the bigger picture and all.

How would you define the bigger picture? The multilingual feature should provide a frame. A more precise context about parts that need some thoughts could help.

I don't want to rush by adding this as a last-minute feature if we release the next version of Hugo soon. It's just a nice addition that fills a gap.

moorereason commented:
I don't like the `strftime` formatting.  That's the bigger picture I'm concerned about.  If this PR used Go time formats, I'd be on-board.

bep commented:
>I don't like the strftime formatting. That's the bigger picture I'm concerned about.

That is an extremely narrow minded picture ... I don't see the bigger picture, but I'm hoping someone else, specialists outside this project, would paint it for me. We're in the process of pulling lots of small localize* into a fairly fragmented and hard to maintain set of functions; there are language rulesets, dictionaries etc. duplicated, most likely not complete (I notice some shortcomings in my own language) ... 

We're not adding this in 0.17. What gets added is very hard to remove.

variadico commented:
@moorereason, Go makes international dates literally impossible. That's why `lctime` exists.
https://github.com/variadico/lctime#the-problem-with-the-go-standard-library

This is a trade off between slightly more idiomatic code—and giving non-English users the luxury of dates.
